### PR TITLE
AFR-614 Automatically remove <> on URLs

### LINKED ISSUE #1228: AFR-614 Automatically remove `<>` on URLs
**Feature Request:** Automatically remove `<>` on URLs
**Who would use it** People who don't quite understand how to read avrae's help syntax
**How would it work** Avrae removes leading `<` and trailing `>` on URLs
**Why should this be added** People mess this up a lot, and it would be a nice QoL thing

- silverbass#2407

> Upvote - Infinidoge#1337

> Downvote - Lathaon#6649

> Upvote - spamandtuna#9657

> Upvote - Riptidecord#7191

> Downvote - Aether_Blackbolt#1361

> Upvote - Swiftgale#4167

> Upvote - Zekl#1505

> Upvote - silverbass#2407

> Upvote - R to the Ichie#6193

> Upvote - Nerds and Dragons#2817

> Upvote - DavidB091#5310

> Upvote - TheReverendB#1377

> Upvote - Shayce#7843

> Downvote - Robbe#7964

> Upvote - Dsg#7515

> Upvote - Retr0#5626

> Upvote - Star☆#2456

> Upvote - Agathist#1706

Votes: +15 / -3

### PR DESCRIPTION
**Pull request by ChrisHamilton91**
### Summary
Resolves #1228
I had someone adding brackets around their url and discord was hiding them from view, making it hard to tell what was wrong. Seems like a pretty easy user error to avoid, we just allow brackets around the url. I'm told I can't test it myself since I won't have access to D&D beyond with a local copy, so hopefully you guys have a unit test or something.
#### PR Type

- [x] This PR is a code change that implements a feature request.
- [ ] This PR fixes an issue.
- [ ] This PR adds a new feature that is not an open feature request.
- [ ] This PR is not a code change (e.g. documentation, README, ...)
#### Other
- [ ] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [x] If code changes were made then they have been tested.
- [ ] I have updated the documentation to reflect the changes.

### CODE REVIEW

**Comment by storytellermahkasad** [comment]
I would almost like to include a message as well because this reinforces the incorrect assumption that <> are part of the command rather than indicating something to be replaced. Would be simple enough to detect the issue and respond with information in addition\rather than silently correcting.

**Comment by ChrisHamilton91** [comment]
Yeah, that's a good idea. I think the main issue there is that the commands page doesn't have a legend at the top, explaining the command syntax. ie. <> indicate a mandatory argument, | means OR, explain what an optional argument is etc. Seems like most of the documentation doesn't take into account most users are not programmers tbh.

**Comment by 1drturtle** [comment]
It's shown at the bottom of every help command.
image
But yeah, it could be added to the top of <https://avrae.io/commands> to not include brackets in commands and to use quotes around multi-word arguments.

**Comment by ChrisHamilton91** [comment]
> It's shown at the bottom of every help command.
> image
> But yeah, it could be added to the top of https://avrae.io/commands to not include brackets in commands and to use quotes around multi-word arguments.

Yeah that's good, but it's not really a full explanation that allows a non-programmer to read the commands page. The people I play with don't even understand where to put whitespace lol. Things that seem obvious to us are not so obvious for everyone.

**Comment by ChrisHamilton91** [comment]
And can anyone explain to me why main arguments don't have descriptions on the commands page? Seems like someone meant to fill them in and just never did.
arguments

**Comment by zhudotexe** [comment]
I'm open to suggestions as to how better to word the note about brackets, but the issue linked in this PR was closed as I don't think silently modifying user input is the right thing to do, and it reinforces bad habits when users are starting to learn the bot. I'd also accept detecting brackets and raising an error instead of silently accepting them.

**Comment by zhudotexe** [comment]
> And can anyone explain to me why main arguments don't have descriptions on the commands page? Seems like someone meant to fill them in and just never did.
> arguments

Remnants of a documentation rewrite that never happened. I'll remove them.

**Comment by ChrisHamilton91** [comment]
> I'm open to suggestions as to how better to word the note about brackets, but the issue linked in this PR was closed as I don't think silently modifying user input is the right thing to do, and it reinforces bad habits when users are starting to learn the bot. I'd also accept detecting brackets and raising an error instead of silently accepting them.

I was thinking of still accepting them, but raising a message like:
"Hey, you put '<' or '>' in your url. Avrae removed them, but don't do that."
I think that is the most user friendly solution.
Reason being that the current error message such as "This is not a D&D beyond link" is not helpful to the user in this situation.
And forcing the user to repeat an action, when we know their intent - I personally don't believe that's a good design practice. I think user experience should be top priority.

On a side note @zhudotexe what's the best way to go about testing? I'm pretty knew to python projects and open source projects in general. As in this is my first PR to an open source project ever...

**Comment by ChrisHamilton91** [comment]
A few other reasons why I think the change is important:

- This is generally a new user's first command with the bot, having it not work right away is a bad first impression
- Brand new users don't necessarily know about the help command
- Discord makes the brackets disappear, so even someone experienced with the bot might not know what's wrong - this one's pretty important since in most cases people are learning from other people, not from the help menu.

**Comment by zhudotexe** [comment]
> I was thinking of still accepting them, but raising a message like:
"Hey, you put '<' or '>' in your url. Avrae removed them, but don't do that."

I'd be fine with this approach, as long as it's not a silent modification of the input. As for testing, you should be able to import Dicecloud/Google sheets if you set Avrae up locally, and I'll test DDB import (sheet import is tricky to test automatically).

**Comment by ChrisHamilton91** [comment]
> 
> 
> > I was thinking of still accepting them, but raising a message like:
> > "Hey, you put '<' or '>' in your url. Avrae removed them, but don't do that."
> 
> I'd be fine with this approach, as long as it's not a silent modification of the input. As for testing, you should be able to import Dicecloud/Google sheets if you set Avrae up locally, and I'll test DDB import (sheet import is tricky to test automatically).

Awesome, does this look good to you?
```python
@staticmethod
async def _check_url(ctx, url):
      if any([c in url for c in ['<', '>']]):
          url = url.strip('<>')
          await ctx.send("Hey! You put '<' or '>' in your url. Avrae removed them, but don't do that!"
                         f"Use `{ctx.prefix}help` for more details")
      return url
```

**Comment by ChrisHamilton91** [comment]
Testing:
dicecloud
gsheet
beyond

File: cogs5e/sheetManager.py [hunk 1]
```diff
@@ -593,6 +596,14 @@ async def _load_sheet(ctx, parser, args, loading):
         await ctx.send(embed=character.get_sheet_embed())
         return character
 
+    @staticmethod
+    async def _check_url(ctx, url):
+        if any([c in url for c in '<>']):
```
Comment (zhudotexe) [inline cogs5e/sheetManager.py:601]:
```suggestion
        if url.startswith('<') and url.endswith('>'):
```

File: cogs5e/sheetManager.py [hunk 2]
```diff
@@ -593,6 +596,14 @@ async def _load_sheet(ctx, parser, args, loading):
...
+            url = url.strip('<>')
+            await ctx.send("Hey! You put '<' or '>' in your url. Avrae removed them, but don't do that!"
```
Comment (zhudotexe) [inline cogs5e/sheetManager.py:603]:
```suggestion
            await ctx.send("Hey! Looks like you surrounded that URL with '<' and '>'. I removed them, but remember not to include those for other arguments!"
```

**Pull request merged**